### DISCUSSION TITLE
Recursive cycles

### DISCUSSION THREAD

**Discussion by voidpetal** [LukasZahradnik/PyNeuraLogic#45] [General]
Hello,

I am trying to write a recursive rule which would tell me if two nodes are in the same cycle of arbitrary length. This is needed for chemical datasets, which have predicates `R.bond(V.X, V.Y, V.Bond)` defined, and it is symmetric. The problem with the standard recursive approach is that in that case every node would be in the same cycle. Ideally, I could pass all the bonds encountered so far, and ask that the next bond is different from all the previous ones, but I would need lists for that, which as far as I understand are not supported.

I would appreciate if you could give me any advice on how to implement this.

Thank you!

**Comment by LukasZahradnik** [comment]
Hi,

Currently, this is not possible for a cycle of arbitrary length. You could define a rule for each cycle length, but that will not be generic.

Lists are indeed not supported, and I think it would require a lot of work to implement them. There is a way to fake them that doesn't require that much work. It is related to the issue you opened before.

@GustikS Should we roll out the experimental join predicate? We could add another predicate for testing if a term is in such joined value. Its performance will be bad (because the domain integer set is fixed size, and we cannot simply expand it), but better than nothing?https://github.com/GustikS/NeuraLogic/compare/master...LukasZahradnik:NeuraLogic:experimental-domain-expansion

It will also have an impact on the whole grounding process because there is a check on whether a rule contains the join predicate, but that should be negligible.

**Comment by GustikS** [comment]
This is of course a very reasonable request and a very nice idea on how to implement it, but I would be careful with changing the grounder at the moment, until we have the (historical) performance testing/logging workflow set up properly...

@erhc For now, and generally for the molecular data, I'd recommend simply listing out the valid cycle lengths - it is very few (5&6?), it will be faster, and generally more efficient by all means...

But it is good that we know about this, and should roll it out once we're ready I guess.

**Comment by voidpetal** [comment]
Okay, I am implementing these cycles manually, and I arrived at another problem. When using negations along with the cycles, it somehow produces the following error when building the dataset:

![Screenshot_1](https://user-images.githubusercontent.com/37145438/221562915-3fc2fc2a-6f01-458e-91e6-e233031cb0be.png)

This is my code (rather long because of cycles of different length):
```
_, dataset = Mutagenesis()

template = Template()

template.add_rules([
    (R.atom_embed(V.A)[3,] <= R.get(atom)(V.A)) for atom in ["c", "o", "br", "i", "f", "h", "n", "cl"]
])


template.add_rules([
    (R.bond_embed(V.B)[3,] <= R.get(bond)(V.B)) for bond in ["b_1", "b_2", "b_3", "b_4", "b_5", "b_7"]
])

template += R.cycle(V.A, V.B) <= (R.cycle(V.A, V.B, V.C))

# 3
template += R.cycle(V.A, V.B, V.C) <= (R.atom_embed(V.A)[3,3],
                                       R.atom_embed(V.B)[3,3],
                                       R.atom_embed(V.C)[3,3],
                                       R.bond_embed(V.B1)[3,3],
                                       R.bond_embed(V.B2)[3,3],
                                       R.bond_embed(V.B3)[3,3],

                                       R.bond(V.A, V.B, V.B1),
                                       R.bond(V.B, V.C, V.B2),
                                       R.bond(V.C, V.A, V.B3),
                                       R.special.alldiff(...))
template += R.cycle(V.A, V.B, V.C) <= (R.cycle(V.A, V.B, V.C, V.D))

# 4
template += R.cycle(V.A, V.B, V.C, V.D) <= (R.atom_embed(V.A)[3,3],
                                            R.atom_embed(V.B)[3,3],
                                            R.atom_embed(V.C)[3,3],
                                            R.atom_embed(V.D)[3,3],
                                            R.bond_embed(V.B1)[3,3],
                                            R.bond_embed(V.B2)[3,3],
                                            R.bond_embed(V.B3)[3,3],
                                            R.bond_embed(V.B4)[3,3],

                                            R.bond(V.A, V.B, V.B1),
                                            R.bond(V.B, V.C, V.B2),
                                            R.bond(V.C, V.D, V.B3),
                                            R.bond(V.D, V.A, V.B4),
                                            R.special.alldiff(...))
template += R.cycle(V.A, V.B, V.C, V.D) <= (R.cycle(V.A, V.B, V.C, V.D, V.E))

# 5
template += R.cycle(V.A, V.B, V.C, V.D, V.E) <= (R.atom_embed(V.A)[3,3],
                                                 R.atom_embed(V.B)[3,3],
                                                 R.atom_embed(V.C)[3,3],
                                                 R.atom_embed(V.D)[3,3],
                                                 R.atom_embed(V.E)[3,3],
                                                 R.bond_embed(V.B1)[3,3],
                                                 R.bond_embed(V.B2)[3,3],
                                                 R.bond_embed(V.B3)[3,3],
                                                 R.bond_embed(V.B4)[3,3],
                                                 R.bond_embed(V.B5)[3,3],

                                                 R.bond(V.A, V.B, V.B1),
                                                 R.bond(V.B, V.C, V.B2),
                                                 R.bond(V.C, V.D, V.B3),
                                                 R.bond(V.D, V.E, V.B4),
                                                 R.bond(V.E, V.A, V.B5),
                                                 R.special.alldiff(...))
template += R.cycle(V.A, V.B, V.C, V.D, V.E) <= (R.cycle(V.A, V.B, V.C, V.D, V.E, V.F))

# 6
template += R.cycle(V.A, V.B, V.C, V.D, V.E, V.F) <= (R.atom_embed(V.A)[3,3],
                                                      R.atom_embed(V.B)[3,3],
                                                      R.atom_embed(V.C)[3,3],
                                                      R.atom_embed(V.D)[3,3],
                                                      R.atom_embed(V.E)[3,3],
                                                      R.atom_embed(V.F)[3,3],
                                                      R.bond_embed(V.B1)[3,3],
                                                      R.bond_embed(V.B2)[3,3],
                                                      R.bond_embed(V.B3)[3,3],
                                                      R.bond_embed(V.B4)[3,3],
                                                      R.bond_embed(V.B5)[3,3],
                                                      R.bond_embed(V.B6)[3,3],
                                                      
                                                      R.bond(V.A, V.B, V.B1),
                                                      R.bond(V.B, V.C, V.B2),
                                                      R.bond(V.C, V.D, V.B3),
                                                      R.bond(V.D, V.E, V.B4),
                                                      R.bond(V.E, V.F, V.B5),
                                                      R.bond(V.F, V.A, V.B6),
                                                      R.special.alldiff(...))
template += R.cycle(V.A, V.B, V.C, V.D, V.E, V.F) <= (R.cycle(V.A, V.B, V.C, V.D, V.E, V.F, V.G))

template += R.predict[1, 3] <= (R.cycle(V.X, V.C)[3, 3], R.c(V.C), ~R._c(V.X))
template += R.predict[1, 3] <= R.cycle(V.X, V.Y)[3,3]
```
It produces the same error in this case: `template += R.predict[1, 3] <= (R.cycle(V.X, V.C)[3, 3], R.c(V.C), ~R.c(V.X))`

I am putting this here and not in the discussion #40, because it seems to work when I replace the `predict` predicate by:
`template += R.predict[1, 3] <= (R.bond_embed(V.B)[3, 3], R.bond(V.X, V.C, V.B), R.c(V.C), ~R.c(V.X)) # it also works with ~R._c(V.X)` 

I would appreciate your help on this, thank you!

**Comment by LukasZahradnik** [comment]
Hi, what is going on is that once you set a relation as hidden (even implicitly via setting it as negative), it will be set as hidden globally.
I think this is expected behavior, although it does not seem to be intuitive and doesn't fit every use case.

@GustikS Would it be possible to include hidden/special in keys of [WeightedPredicateFactory](https://github.com/GustikS/NeuraLogic/blob/master/Logical/src/main/java/cz/cvut/fel/ida/logic/constructs/building/factories/WeightedPredicateFactory.java) to fix this, or will it break anything?

Btw, you can generate those loops, which would simplify your code a lot, for example:
```py
def get_cycle(n):
    body = [R.bond(f"X{i}", f"X{(i+1) % n}", f"B{i}") for i in range(n)]
    body.extend(R.atom_embed(f"X{i}")[3, 3] for i in range(n))
    body.extend(R.bond_embed(f"B{i}")[3, 3] for i in range(n))
    body.append(R.special.alldiff(...))

    return R.cycle([f"X{i}" for i in range(n)]) <= body

template += get_cycle(3)
# cycle(X0, X1, X2) :- bond(X0, X1, B0), bond(X1, X2, B1), bond(X2, X0, B2), {3, 3} atom_embed(X0), {3, 3} atom_embed(X1), {3, 3} atom_embed(X2), {3, 3} bond_embed(B0), {3, 3} bond_embed(B1), {3, 3} bond_embed(B2), @alldiff(B1, B0, B2, X2, X0, X1).
...


```

**Reply by GustikS** [reply]
Hi, I agree that setting to hidden/special explicitly should be global (it's a feature of the relation), but implicitly through the negation is kind of an undesired side effect I'm afraid (negation is not a feature of the relation) - could that be resolved by setting the negation to the default value instead of hidden as discussed in https://github.com/LukasZahradnik/PyNeuraLogic/discussions/40? 

+1 for the loops

**Reply by LukasZahradnik** [reply]
@GustikS Hi, it should be resolved by setting the negation to the default value. Although, wouldn't it be better to also allow having the hidden and not hidden relation in a template? It seems kinda weird that hiding a relation in one rule would affect the behavior of a completely different rule.

**Reply by GustikS** [reply]
True, I was kinda just proposing an "easy" fix of this specific issue, but generally I can also imagine the hidden to work on the level of atoms instead of predicates, but that would require quite a few changes in the backend I guess :) In majority of cases it will be global I'd say, while negation is clearly local imho...

**Reply by voidpetal** [reply]
Update: I solved this issue by introducing a new predicate `R.n_c(V.X) <= R.c(V.X)` and then negating it. However, it still would be neat to be able to put the same predicate both negated and non-negated in the same body.
